pkg/server/endpoints/endpoints.go

```diff
@@ -0,0 +1,359 @@
+	"github.com/spiffe/spire/pkg/server/endpoints/registration"
+
+	node_pb "github.com/spiffe/spire/proto/api/node"
+	registration_pb "github.com/spiffe/spire/proto/api/registration"
+	ca_pb "github.com/spiffe/spire/proto/server/ca"
+
+	"golang.org/x/net/context"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"gopkg.in/tomb.v2"
+)
+
+// Server manages gRPC and HTTP endpoint lifecycle
+type Server interface {
+	// ListenAndServe starts all endpoints, and blocks for as long as the
+	// underlying servers are still running. Returns an error if any of the
+	// endpoints encounter one.
+	ListenAndServe() error
+
+	// Shutdown gracefully closes all underlying endpoint servers.
+	// ListenAndServe will unblock with `nil` if/when shutdown completes
+	// cleanly
+	Shutdown()
+}
+
+type endpoints struct {
+	c   *Config
+	mtx *sync.RWMutex
+	t   *tomb.Tomb
+
+	svid    *x509.Certificate
+	svidKey *ecdsa.PrivateKey
+	caCerts []*x509.Certificate
+
+	svidCheck *time.Ticker
+
+	grpcServer *grpc.Server
+	httpServer *http.Server
+
+	runOnce *sync.Once
+}
+
+// ListenAndServe starts all maintenance routines and endpoints, then blocks
+// for as long as the underlying servers are still running. Returns an error
+// if any of the endpoints encounter one.
+func (e *endpoints) ListenAndServe() error {
+	run := func() { e.t.Go(e.listenAndServe) }
+	e.runOnce.Do(run)
+
+	return e.t.Wait()
+}
+
+// Shutdown gracefully closes all underlying endpoint servers.
+func (e *endpoints) Shutdown() {
+	e.t.Kill(nil)
+	return
+}
+
+// listenAndServe creates listeners and starts all servers. It serves
+// as the top-most tomb routine.
+func (e *endpoints) listenAndServe() error {
+	// Certs must be ready before anything else
+	err := e.rotateSVID()
+	if err != nil {
+		return err
+	}
+
+	e.c.Log.Debug("Initializing API endpoints")
+	gs := e.createGRPCServer()
+	hs := e.createHTTPServer()
+
+	e.registerNodeAPI(gs)
+	err = e.registerRegistrationAPI(gs, hs)
+	if err != nil {
+		return err
+	}
+
+	e.grpcServer = gs
+	e.httpServer = hs
+
+	e.t.Go(e.startGRPCServer)
+	e.t.Go(e.startHTTPServer)
+	e.t.Go(e.startRotator)
+
+	return nil
+}
+
+func (e *endpoints) createGRPCServer() *grpc.Server {
+	e.mtx.RLock()
+	defer e.mtx.RUnlock()
+
+	caPool := x509.NewCertPool()
+	for _, c := range e.caCerts {
+		caPool.AddCert(c)
+	}
+
+	tlsConfig := &tls.Config{
+		// When bootstrapping, the agent does not yet have
+		// an SVID. In order to include the bootstrap endpoint
+		// in the same server as the rest of the Node API,
+		// request but don't require a client certificate
+		ClientAuth: tls.RequestClientCert,
+
+		// Hook TLS with a custom function so we can always provide the
+		// latest TLS credentials.
+		GetCertificate: e.getSVID,
+
+		// Use our own CA certs as the root for client auth
+		ClientCAs: caPool,
+	}
+
+	opts := grpc.Creds(credentials.NewTLS(tlsConfig))
+	return grpc.NewServer(opts)
+}
+
+func (e *endpoints) createHTTPServer() *http.Server {
+	e.mtx.RLock()
```

Comment:
Heh yea... It was the only method like it without a read lock (because it didn't need one, as you pointed out)... then I thought, if it's ever changed in the future adding the lock will be easy to overlook, and when reviewing the code the seemingly missing lock invoked some thought on my part, so I decided to add it even though we didn't actually need it.

I'm open for debate on this point of course :) if folks feel that is a silly reason to keep it, I'm happy to remove it.